### PR TITLE
Word wrapping

### PR DESCRIPTION
**Pull request by karimnaaji**
as in [tangram-js](https://github.com/tangrams/tangram/pull/205)

- [x] alignments (center/left/right)
- [x] add stylesheet support

screen shot 2015-10-27 at 10 23 17 am

### CODE REVIEW

File: core/src/text/textBuffer.cpp [hunk 1]
```diff
@@ -15,6 +15,22 @@ TextBuffer::TextBuffer(std::shared_ptr<VertexLayout> _vertexLayout)
 TextBuffer::~TextBuffer() {
 }
 
+std::vector<TextBuffer::WordBreak> TextBuffer::findWordBreaks(const std::string& _text) {
+    std::vector<WordBreak> breaks;
+
+    for (int i = 0, quad = 0; i < _text.size(); ++i) {
+        if (std::isspace(_text[i])) {
+            int lookAHead = i;
```
Comment (bcamper) [inline core/src/text/textBuffer.cpp]: very minor nitpick, the H in lookAHead is probably not meant to be capitalized :)

File: core/src/text/textBuffer.cpp [hunk 2]
```diff
@@ -15,6 +15,22 @@ TextBuffer::TextBuffer(std::shared_ptr<VertexLayout> _vertexLayout)
 TextBuffer::~TextBuffer() {
 }
 
+std::vector<TextBuffer::WordBreak> TextBuffer::findWordBreaks(const std::string& _text) {
```
Comment (bcamper) [inline core/src/text/textBuffer.cpp]: Looks like we also need to handle explicit line breaks (`\n` chars) in the text, to match JS (unless I missed it somewhere). For example, the string `One\nTwo\nThree` should render as:

```
One
Two
Three
```

File: core/src/text/textBuffer.cpp (hunk 1 above)
Comment (karimnaaji) [inline core/src/text/textBuffer.cpp]: Oh yes :)

File: core/src/text/textBuffer.cpp (hunk 2 above)
Comment (karimnaaji) [inline core/src/text/textBuffer.cpp]: Line breaks `\n` is something that can come from the string data?

Comment (bcamper) [inline core/src/text/textBuffer.cpp]: Yep, either if it was in the source data, or added in the stylesheet by a JS function, e.g. `text_source: function(){ return feature.name.replace(' ', '\n') }`. When there are explicit line breaks, we assume that they were desired, so we always break on them.

Comment (karimnaaji) [inline core/src/text/textBuffer.cpp]: Ok makes sense, I didn't assume that, but will update considering this.

File: core/src/labels/textLabel.cpp [hunk 3]
```diff
@@ -20,9 +21,10 @@ void TextLabel::updateBBoxes(float _zoomFract) {
     obbCenter = m_transform.state.screenPos;
     // move forward on line by half the text length
     obbCenter += t * m_dim.x * 0.5f;
-    // TODO: use real font metrics
     // move down on the perpendicular to estimated font baseline
-    obbCenter -= tperp * (m_dim.y / 8);
+    obbCenter -= tperp * m_dim.y * 0.5f;
+    // ajdust with baseline
+    obbCenter.y += m_metrics.lineHeight * (float) m_nLines * 0.5f;
```
Comment (tallytalwar) [inline core/src/labels/textLabel.cpp]: Don't think we need to have `float` typecasting here.

File: core/src/text/fontContext.h [hunk 4]
```diff
@@ -64,10 +73,12 @@ class FontContext {
     void initFontContext(int _atlasSize);
 
     std::map<std::string, int> m_fonts;
+    std::map<int, FontMetrics> m_fontMetrics;
```
Comment (tallytalwar) [inline core/src/text/fontContext.h]: I don't see this being used anywhere.

**Comment by tallytalwar** [comment]
Missing `word_wrap` parameter parsing to have styling control for text wrapping. Currently only default values will be respected.

**Comment by bcamper** [comment]
@tallytalwar it's actually `text_wrap`, but it looks like @karimnaaji got it.

**Comment by tallytalwar** [comment]
@bcamper yup. Sorry meant that :D.

**Comment by tallytalwar** [comment]
This looks good to me now.

**Pull request merged**